### PR TITLE
validation and mutating webhook for experiment

### PR DESCRIPTION
**Pull request by hougangliu**
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/katib/473)

### CODE REVIEW

File: pkg/api/operators/apis/experiment/v1alpha2/experiment_defaults.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+// Code generated by main. DO NOT EDIT.
+
+package v1alpha2
+
+func (e *Experiment) Default() {
```
Comment (johnugeorge) [inline pkg/api/operators/apis/experiment/v1alpha2/experiment_defaults.go]: Many be SetDefaults?

File: pkg/controller/v1alpha2/experiment/util/webhook_util.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	if len(instance.Spec.Parameters) == 0 && instance.Spec.NasConfig == nil {
+		return fmt.Errorf("spec.Parameters or spec.NasConfig must be specified.")
+	}
+
+	if len(instance.Spec.Parameters) > 0 && instance.Spec.NasConfig != nil {
+		return fmt.Errorf("Only one of spec.Parameters and spec.NasConfig can be specified.")
+	}
+
+	if len(instance.Spec.Parameters) > 0 {
+		if err := validateParameters(instance.Spec.Parameters); err != nil {
```
Comment (andreyvelich) [inline pkg/controller/v1alpha2/experiment/util/webhook_util.go]: Will we use `ValidateAlgorithmSetting` function to validate parameters inside Suggestion ? 
https://github.com/kubeflow/katib/blob/master/pkg/api/v1alpha2/api.proto#L170

Comment (hougangliu) [reply pkg/controller/v1alpha2/experiment/util/webhook_util.go]: sure :)

Comment (andreyvelich) [reply pkg/controller/v1alpha2/experiment/util/webhook_util.go]: Good, thank you!

File: pkg/api/operators/apis/experiment/v1alpha2/experiment_defaults.go (hunk 1 above)
Comment (hougangliu) [reply pkg/api/operators/apis/experiment/v1alpha2/experiment_defaults.go]: I can update it (in fact, I see many kube project use Default(), too).

**Comment by johnugeorge** [comment]
@hougangliu Can you have a look at my previous comment?  Is it possible to avoid client-go dependency directly?  The problem is that we will then have to keep it as a direct dependency in Godep and then have to worry about the changes in client-go library. However, if we use client of controller-runtime(iwhich indirectly uses client-go), we have to just upgrade controller-runtime whenever required

**Comment by hougangliu** [comment]
> @hougangliu Can you have a look at my previous comment? Is it possible to avoid client-go dependency directly? The problem is that we will then have to keep it as a direct dependency in Godep and then have to worry about the changes in client-go library. However, if we use client of controller-runtime(iwhich indirectly uses client-go), we have to just upgrade controller-runtime whenever required

OK, I can try. Thanks!

**Comment by johnugeorge** [comment]
Sorry. Forgot to publish

File: pkg/controller/v1alpha2/experiment/util/template_util.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	"bytes"
+	"errors"
+	"os"
+	"text/template"
+
+	apiv1alpha2 "github.com/kubeflow/katib/pkg/api/v1alpha2"
+	experimentsv1alpha2 "github.com/kubeflow/katib/pkg/api/operators/apis/experiment/v1alpha2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
```
[the remark this answers is not in the document]
Comment (hougangliu) [reply pkg/controller/v1alpha2/experiment/util/template_util.go]: done! thanks

File: pkg/api/operators/apis/experiment/v1alpha2/constants.go [hunk 4]
```diff
@@ -27,4 +27,7 @@ const (
 
 	// Default value of Spec.TemplatePath
 	DefaultTrialTemplatePath = "defaultTrialTemplate.yaml"
+
+	OptimizationTypeMinimize = "minimize"
```
Comment (johnugeorge) [inline pkg/api/operators/apis/experiment/v1alpha2/constants.go]: Do we need this? Isn't it already defined in types.go?

Comment (hougangliu) [reply pkg/api/operators/apis/experiment/v1alpha2/constants.go]: I will remove it

File: pkg/controller/v1alpha2/experiment/util/template_util.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+	trialTemplate := instance.Spec.TrialTemplate
+	if trialTemplate != nil && trialTemplate.GoTemplate.RawTemplate != "" {
+		tpl, err = template.New("Trial").Parse(trialTemplate.GoTemplate.RawTemplate)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		configMapNS := os.Getenv(experimentsv1alpha2.DefaultKatibNamespaceEnvName)
+		configMapName := experimentsv1alpha2.DefaultTrialConfigMapName
```
Comment (johnugeorge) [inline pkg/controller/v1alpha2/experiment/util/template_util.go]: Should we keep these too in the defaults?

Comment (hougangliu) [reply pkg/controller/v1alpha2/experiment/util/template_util.go]: make sense

File: pkg/controller/v1alpha2/experiment/util/webhook_util.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+import (
+	"bytes"
+	"fmt"
+
+	ep_v1alpha2 "github.com/kubeflow/katib/pkg/api/operators/apis/experiment/v1alpha2"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	k8syaml "k8s.io/apimachinery/pkg/util/yaml"
+)
+
+func ValidateExperiment(instance *ep_v1alpha2.Experiment) error {
```
Comment (YujiOshima) [inline pkg/controller/v1alpha2/experiment/util/webhook_util.go]: Could you add validation for Experiment name that the name is already registered in DB or not?
If users set `RetainHistoricalData` true, the previous data will remain even the CR was deleted.

Comment (hougangliu) [reply pkg/controller/v1alpha2/experiment/util/webhook_util.go]: done! thanks!

**Pull request merged**